lib/fields.c

```diff
@@ -84,15 +84,15 @@ void change_field (char *buf, size_t maxsize, const char *prompt)
 	if (NULL == cp) {
 		return;
 	}
-	*cp = '\0';
+	stpcpy(cp, "");
```

Comment:
I'll merge it, but for the record I really don't see stpcpy(cp, "") as an improvement over *cp = '\0' (here or in the other cases in this commit).  Everyone understands the latter.  And especially in something like:

```
1412 -       name[max] = '\0';
1413 +       stpcpy(&name[max], "");
```

Reply:
The main benefit is _FORTIFY_SOURCE.

I got that habit when you asked me to make sure that the last byte written in one of the string functions that I developed:

<https://github.com/shadow-maint/shadow/pull/850#discussion_r1413169325>

Reply:
> The main benefit is _FORTIFY_SOURCE.
> 
> I got that habit when you asked me to make sure that the last byte written in one of the string functions that I developed was actually within buffer bounds:
> 
> <https://github.com/shadow-maint/shadow/pull/850#discussion_r1413167760>
> <https://github.com/shadow-maint/shadow/pull/850#discussion_r1413169325>

ah, interesting, will read up on that. The direct assignment cannot pass with that flag, or whT? That sounds like a good reason.

Reply:
I'm not an expert, so maybe GCC does some heavier magic, but AFAIK, `_FORTIFY_SOURCE` is implemented as some wrapper macros (or inline functions) that are not much more than our upper-case macros:

```c
alx@debian:~/src/gnu/glibc/master$ sed -n '20,40p' string/bits/string_fortified.h

#ifndef _STRING_H
# error "Never use <bits/string_fortified.h> directly; include <string.h> instead."
#endif

__fortify_function void *
__NTH (memcpy (void *__restrict __dest, const void *__restrict __src,
	       size_t __len))
{
  return __builtin___memcpy_chk (__dest, __src, __len,
				 __glibc_objsize0 (__dest));
}

__fortify_function void *
__NTH (memmove (void *__dest, const void *__src, size_t __len))
{
  return __builtin___memmove_chk (__dest, __src, __len,
				  __glibc_objsize0 (__dest));
}

#ifdef __USE_GNU
```

This is, apart from relying on compiler builtins, not much more magic than our STRNCPY():

```c
$ grepc STRNCPY .
./lib/string/strcpy/strncpy.h:#define STRNCPY(dst, src)  strncpy(dst, src, NITEMS(dst))
```

So, yes, explicit byte copying is not protected.  Only libc functions that internally call these compiler builtins offer protection.